src/autodiff.jl

```diff
@@ -145,6 +145,69 @@ struct Jacobian{VI, VT, MT, SMT, VP, VD, SubT, SubD}
     end
 end
 
+"""
+    Hessian
+
+Creates an object for computing Hessian adjoint tangent projections
+
+* `t1sseeds::VP`: The seeding vector for AD built based on the coloring.
+* `t1sF::VD`: Output array of active (AD) type.
+* `x::VT`: Input array of passive type. This includes both state and control.
+* `t1sx::VD`: Input array of active type.
+* `map::VI`: State and control mapping to array `x`
+* `varx::SubT`: View of `map` on `x`
+* `t1svarx::SubD`: Active (AD) view of `map` on `x`
+"""
+struct Hessian{VI, VT, MT, SMT, VP, VD, SubT, SubD} <: AbstractHessian
+    t1sseeds::VP
+    t1sF::VD
+    x::VT
+    t1sx::VD
+    map::VI
+    # Cache views on x and its dual vector to avoid reallocating on the GPU
+    varx::SubT
+    t1svarx::SubD
+    function Hessian(structure, F, vm, va, ybus_re, ybus_im, pinj, qinj, pv, pq, ref)
+        nvbus = length(vm)
+        npbus = length(pinj)
+        nref = length(ref)
+        if F isa Array
+            VI = Vector{Int}
+            VT = Vector{Float64}
+            MT = Matrix{Float64}
+            SMT = SparseMatrixCSC
+            A = Vector
+        elseif F isa CUDA.CuArray
+            VI = CUDA.CuVector{Int}
+            VT = CUDA.CuVector{Float64}
+            MT = CUDA.CuMatrix{Float64}
+            SMT = CUSPARSE.CuSparseMatrixCSR
+            A = CUDA.CuVector
+        else
+            error("Wrong array type ", typeof(F))
+        end
+
+        map = VI(structure.map)
+        nmap = length(structure.map)
+        # Need a host arrays for the sparsity detection below
+        if F isa CUDA.CuArray
+            J = CUSPARSE.CuSparseMatrixCSR(J)
+        end
+        t1s{N} = ForwardDiff.Dual{Nothing,Float64, N} where N
+        x = VT(zeros(Float64, 2*nvbus + npbus))
+        t1sx = A{t1s{1}}(x)
```

Comment:
I just did a pass on type inference. Here, `t1sx` because `A` is a generic type `Array` (on the CPU), whereas it should be inferred as a `Array{Dual{Nothing, Float64, 1}}`

Reply:
We should look into all of this. Also the seeding. I agree, let's to that in a performance specific PR.